### PR TITLE
Power-off and power-on all worker nodes

### PR DESCRIPTION
**Pull request by Santhoshairody**
Add PowerOffAllVMs to longevity
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** (optional)
Closes #
PTX-23580

**Special notes for your reviewer**
http://jen-endurance.pwx.dev.purestorage.com/job/Users/job/santhoshm/job/IOProfile-update/53/console

### CODE REVIEW

File: tests/testTriggers.go [hunk 1]
```diff
@@ -6458,6 +6459,125 @@ func TriggerVolumeUpdate(contexts *[]*scheduler.Context, recordChan *chan *Event
...
+		var counter int = 0
+		nodesInThread := make([][]node.Node, numberOfThread)
+		for t := 0; t < numberOfThread;  t++ {
+			nodesInThread[t] = make([]node.Node, numberOfNodePerThread)
+			for n:=0; n < numberOfNodePerThread; n++{
+					nodesInThread[t][n] = workerNodes[counter]
+					counter++
+			}
+		}
+		log.Infof("before additional nodes  : %v", nodesInThread)
```
Comment (Rohit-PX) [inline tests/testTriggers.go:6502]: Remove these logs.

File: tests/testTriggers.go [hunk 2]
```diff
@@ -6458,6 +6459,125 @@ func TriggerVolumeUpdate(contexts *[]*scheduler.Context, recordChan *chan *Event
...
+			for counter< len(workerNodes) {
+				additonalThread[index] = workerNodes[counter]
+				index++
+				counter++
+			}
+			nodesInThread = append(nodesInThread, additonalThread)
+			numberOfThread++
+		}
+		log.Infof("Complete nodes  : %v", nodesInThread)
+		log.Infof("Complete nodes len  : %v", len(nodesInThread))
```
Comment (Rohit-PX) [inline tests/testTriggers.go:6518]: Remove both these lines as well.

File: tests/testTriggers.go [hunk 3]
```diff
@@ -6458,6 +6459,125 @@ func TriggerVolumeUpdate(contexts *[]*scheduler.Context, recordChan *chan *Event
...
+	defer func() {
+		event.End = time.Now().Format(time.RFC1123)
+		*recordChan <- event
+	}()
+	setMetrics(*event)
+	stepLog := "Power off all worker nodes test"
+	Step(stepLog, func() {
+		log.Infof(stepLog)
+		workerNodes:= node.GetWorkerNodes()
+		log.Infof("All worker nodes : %v", workerNodes)
```
Comment (Rohit-PX) [inline tests/testTriggers.go:6484]: Remove this line before merging.

File: tests/testTriggers.go [hunk 4]
```diff
@@ -6458,6 +6459,125 @@ func TriggerVolumeUpdate(contexts *[]*scheduler.Context, recordChan *chan *Event
...
+			numberOfNodePerThread = 1
+		}else{
+			numberOfNodePerThread = len(workerNodes)/numberOfThread
+		}
+		var counter int = 0
+		nodesInThread := make([][]node.Node, numberOfThread)
+		for t := 0; t < numberOfThread;  t++ {
+			nodesInThread[t] = make([]node.Node, numberOfNodePerThread)
+			for n:=0; n < numberOfNodePerThread; n++{
+					nodesInThread[t][n] = workerNodes[counter]
```
Comment (Rohit-PX) [inline tests/testTriggers.go:6498]: Need to check if `counter` does not go out of bounds, else we will get index our of range error here.

File: tests/testTriggers.go [hunk 5]
```diff
@@ -6458,6 +6459,125 @@ func TriggerVolumeUpdate(contexts *[]*scheduler.Context, recordChan *chan *Event
...
+			numberOfThread++
+		}
+		log.Infof("Complete nodes  : %v", nodesInThread)
+		log.Infof("Complete nodes len  : %v", len(nodesInThread))
+		stepLog = "Power off all worker nodes"
+		Step(stepLog, func() {
+			var poweroffwg sync.WaitGroup
+			for i := 0; i < numberOfThread; i++ {
+				poweroffwg.Add(1)
+				go func(nodeList []node.Node) {
```
Comment (Rohit-PX) [inline tests/testTriggers.go:6524]: Is the test case designed to poweroff multiple VMs at once?

File: tests/testTriggers.go [hunk 6]
```diff
@@ -6458,6 +6459,125 @@ func TriggerVolumeUpdate(contexts *[]*scheduler.Context, recordChan *chan *Event
...
+							log.Infof("err %v",err)
+					}
+				}(nodesInThread[i])
+			}
+			poweroffwg.Wait()
+			log.Infof("Completed power off VMs")
+			log.Infof("Wait for 5 minutes")
+			time.Sleep(time.Duration(5 * time.Minute))
+		})
+		stepLog = "Power on all worker nodes"
```
Comment (Rohit-PX) [inline tests/testTriggers.go:6539]:
```suggestion
		stepLog = "Power on all worker nodes in batches"
```

File: tests/testTriggers.go [hunk 7]
```diff
@@ -6458,6 +6459,125 @@ func TriggerVolumeUpdate(contexts *[]*scheduler.Context, recordChan *chan *Event
...
+				additonalThread[index] = workerNodes[counter]
+				index++
+				counter++
+			}
+			nodesInThread = append(nodesInThread, additonalThread)
+			numberOfThread++
+		}
+		log.Infof("Complete nodes  : %v", nodesInThread)
+		log.Infof("Complete nodes len  : %v", len(nodesInThread))
+		stepLog = "Power off all worker nodes"
```
Comment (Rohit-PX) [inline tests/testTriggers.go:6519]:
```suggestion
		stepLog = "Power off all worker nodes in batches" 
```

**Review by Rohit-PX — CHANGES_REQUESTED** [review]
Few comments.

File: tests/testTriggers.go [hunk 8]
```diff
@@ -6458,6 +6459,125 @@ func TriggerVolumeUpdate(contexts *[]*scheduler.Context, recordChan *chan *Event
...
+		if len(workerNodes) < numberOfThread{
+			numberOfThread = len(workerNodes)
+			numberOfNodePerThread = 1
+		}else{
+			numberOfNodePerThread = len(workerNodes)/numberOfThread
+		}
+		var counter int = 0
+		nodesInThread := make([][]node.Node, numberOfThread)
+		for t := 0; t < numberOfThread;  t++ {
+			nodesInThread[t] = make([]node.Node, numberOfNodePerThread)
```
Comment (lsrinivas-pure) [inline tests/testTriggers.go:6496]: can you add a comment explaining the logic

File: tests/testTriggers.go (hunk 5 above)
Comment (Santhoshairody) [reply tests/testTriggers.go:6524]: > Is the test case designed to poweroff multiple VMs at once?

yes

File: tests/testTriggers.go (hunk 4 above)
Comment (Santhoshairody) [reply tests/testTriggers.go:6526]: Yes. I checked with different values

**Review by Rohit-PX — APPROVED** [review]
LGTM.

**Comment by Rohit-PX** [comment]
@Santhoshairody please rebase and merge.

**Pull request merged**